com.unity.sg2/Tests/HeadlessPreview/AllNodesTest.cs

```diff
@@ -0,0 +1,134 @@
+using System;
+using NUnit.Framework;
+using System.Collections.Generic;
+using UnityEngine;
+using UnityEditor.ShaderGraph.GraphDelta;
+using System.Linq;
+using UnityEditor.ShaderGraph.HeadlessPreview.UnitTests;
+using UnityEngine.TestTools;
+using static UnityEditor.VersionControl.Message;
+
+namespace UnityEditor.ShaderGraph.HeadlessPreview.NodeTests
+{
+
+    // TODO: Consider async compile/rendering, need to wait until thread is done and results are returned
+    // https://forum.unity.com/threads/support-for-async-await-in-tests.787853/#post-6838493
+
+    [TestFixture]
+    class DefaultPreviewsTestFixture
+    {
+        static ShaderGraphRegistry InitSGR()
+        {
+            ShaderGraphRegistry SGR = new();
+            SGR.InitializeDefaults();
+            return SGR;
+        }
+
+        static string[] InitNodeNames()
+        {
+            return SGR.DefaultTopologies.GetNodes().Select(e => e.ID.LocalPath).ToArray();
+        }
+
+        // whether or not a node will require another node to coerce the value to a vector for testing preview compilation.
+        static bool HasVectorOutput(NodeHandler node)
+        {
+            var nodeKey = node.GetRegistryKey();
+            var typeField = node.GetPorts().Where(e => !e.IsInput)?.First()?.GetTypeField();
+            bool hasPreviewableOutputPort = typeField?.GetRegistryKey().Name == GraphType.kRegistryKey.Name;
+            return hasPreviewableOutputPort && GraphTypeHelpers.GetHeight(typeField) == GraphType.Height.One;
+        }
+
+        static ShaderGraphRegistry SGR = InitSGR();
```

Comment:
I'm not sure what you mean.

Reply:
As is, `SGR` can be overwritten. Do you intend for this to be a singleton, such that each time `SGR` is accessed, if it is empty then a SGR is created and if it already exists the same one is returned? 

If so, then using a static singleton here might be appropriate.

If not, the general recommendation for this seems to be static constructors https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/classes-and-structs/static-constructors However, I'm not sure how this plays in NUnit's lifecycle. I can try this out and get back to this comment thread.

Reply:
After looking over this some more, I think that the registry should be initialized in a method with the `[OneTimeSetup]` NUnit attribute. https://docs.nunit.org/articles/nunit/writing-tests/attributes/onetimesetup.html

What do you think?

Reply:
I started with `[OneTimeSetup]`, but I do not believe it gets called before the `[TestCaseSource("nodeNames")]` populates the test cases in TestRunner. Which is why everything is static here. We wouldn't be able to get _per node_ tests in the list.
--
Oh yea- I know what a singleton is :P.  `ShaderGraphRegistry` has a singleton instance already. As an accessor pattern, it didn't feel relevant for the test. I originally created a local instance to the fixture so that it could be mutated by `[OneTimeSetup]` without impacting the singleton the editor uses, but because of the limitation above I had to make it static to get it to work correctly.